be/src/olap/types.cpp

```diff
@@ -95,110 +47,188 @@ bool is_olap_string_type(FieldType field_type) {
+    case OLAP_FIELD_TYPE_BIGINT:
+        return &bigint_type_info[iterations];
+    case OLAP_FIELD_TYPE_LARGEINT:
+        return &largeint_type_info[iterations];
+    case OLAP_FIELD_TYPE_FLOAT:
+        return &float_type_info[iterations];
+    case OLAP_FIELD_TYPE_DOUBLE:
+        return &double_type_info[iterations];
+    case OLAP_FIELD_TYPE_DECIMAL:
+        return &decimal_type_info[iterations];
+    case OLAP_FIELD_TYPE_DATE:
+        return &date_type_info[iterations];
+    case OLAP_FIELD_TYPE_DATETIME:
+        return &datetime_type_info[iterations];
+    case OLAP_FIELD_TYPE_CHAR:
+        return &char_type_info[iterations];
+    case OLAP_FIELD_TYPE_VARCHAR:
+        return &varchar_type_info[iterations];
+    case OLAP_FIELD_TYPE_STRING:
+        return &string_type_info[iterations];
+    default:
+        DCHECK(false) << "Bad field type: " << leaf_type;
+    }
+    return nullptr;
 }
 
-std::shared_ptr<const TypeInfo> get_type_info(segment_v2::ColumnMetaPB* column_meta_pb) {
-    FieldType type = (FieldType)column_meta_pb->type();
-    if (is_scalar_type(type)) {
-        return get_scalar_type_info(type);
-    } else {
-        switch (type) {
-        case OLAP_FIELD_TYPE_ARRAY: {
-            return ArrayTypeInfoResolver::instance()->get_type_info(*column_meta_pb);
-        }
-        default:
-            DCHECK(false) << "Bad field type: " << type;
-            return nullptr;
+const TypeInfo* get_type_info(segment_v2::ColumnMetaPB* column_meta_pb) {
+    FieldType type = (FieldType) column_meta_pb->type();
+    switch (type) {
+    case OLAP_FIELD_TYPE_ARRAY: {
+        int32_t iterations = 0;
+        const auto* child_column = &column_meta_pb->children_columns(0);
+        while (child_column->type() == OLAP_FIELD_TYPE_ARRAY) {
+            iterations++;
+            child_column = &child_column->children_columns(0);
         }
+        return get_array_type_info((FieldType) child_column->type(), iterations);
+    }
+    [[likely]] default:
+        return get_scalar_type_info(type);
     }
 }
 
-std::shared_ptr<const TypeInfo> get_type_info(const TabletColumn* col) {
-    if (is_scalar_type(col->type())) {
-        return get_scalar_type_info(col->type());
-    } else {
-        switch (col->type()) {
-        case OLAP_FIELD_TYPE_ARRAY:
-            return ArrayTypeInfoResolver::instance()->get_type_info(*col);
-        default:
-            DCHECK(false) << "Bad field type: " << col->type();
-            return nullptr;
+const TypeInfo* get_type_info(const TabletColumn* col) {
+    auto type = col->type();
+    switch (type) {
+    case OLAP_FIELD_TYPE_ARRAY: {
+        int32_t iterations = 0;
+        const auto* child_column = &col->get_sub_column(0);
+        while (child_column->type() == OLAP_FIELD_TYPE_ARRAY) {
+            iterations++;
+            child_column = &child_column->get_sub_column(0);
         }
+        return get_array_type_info(child_column->type(), iterations);
+    }
+    [[likely]] default:
```

Comment:
`[[likely]]` is used for C++20
we can use `UNLIKELY` or  `LIKELY` for C++17 in doris currently

reference: https://en.cppreference.com/w/cpp/language/attributes/likely

Reply:
it seems that we cannot use UNLIKELY or LIKELY for switch now， I will remove it